File: 815-champagne-tower/champagne-tower.cpp
#include <vector>     // For std::vector
#include <numeric>    // For std::iota (not used here, but common for fill)
#include <algorithm>  // For std::max, std::min

class Solution {
public:
    
    std::vector<std::vector<double>> dp;
    int initialPoured; 

    double helper(int r, int c) {
        if (c < 0 || c > r) {
            return 0.0; 
        }

        if (dp[r][c] != -1.0) {
            return dp[r][c];
        }


        if (r == 0 && c == 0) {
            return dp[r][c] = initialPoured;
        }

        double leftOverflow = std::max(0.0, (helper(r - 1, c - 1) - 1.0) / 2.0);
        

        double rightOverflow = std::max(0.0, (helper(r - 1, c) - 1.0) / 2.0);

        return dp[r][c] = leftOverflow + rightOverflow;
    }

    double champagneTower(int poured, int query_row, int query_glass) {
        
        dp.resize(query_row + 1, std::vector<double>(query_row + 1, -1.0));
        this->initialPoured = poured; 

        double totalChampagne = helper(query_row, query_glass);

        return std::min(1.0, totalChampagne);
    }
};

File: 815-champagne-tower/champagne-tower.java
class Solution {
    double[][] dp;
    public double champagneTower(int poured, int query_row, int query_glass) {
        dp = new double[query_row+1][query_row+1];
        for(double[] arr: dp){
            Arrays.fill(arr, -1.0);
        }
        return Math.min(1.0, helper(poured, query_row, query_glass));
    }
     double helper(int poured, int i, int j){
        if(i<0 || j<0 || i<j){
            return 0.0;
        }

        if(dp[i][j] != -1.0){
            return dp[i][j];
        }

        if(i==0 && j==0) return dp[i][j] = poured;

        
        double left_up = Math.max(0, (helper(poured, i - 1, j - 1) - 1) / 2);
        double right_up = Math.max(0, (helper(poured, i - 1, j) - 1) / 2);

        return dp[i][j] = left_up+right_up;
    }
}